### PR TITLE
Don't lint for self-recursion when the function can diverge

### LINKED ISSUE #54444: The `unconditional_recursion` lint can trigger in divergent functions
@eddyb noticed these cases:

```rust
pub fn foo(x: bool) { // WARN function cannot return without recurring
    if x {
        foo(x);
    } else {
        loop {}
    }
}
```
[Try It](https://play.rust-lang.org/?gist=003da5d7fe5cb1fd8ecaefc75fca4fd1&version=stable&mode=debug&edition=2015)

```rust
pub fn foo(x: bool) { // WARN function cannot return without recursing
    if x {
        foo(!x);
    } else {
        panic!("foo");
    }
}
```
[Try It](https://play.rust-lang.org/?gist=e702b4b89d80db107093b36ecc6097c6&version=stable&mode=debug&edition=2015)

@eddyb thinks the way to resolve this is to reimplement the lint on top of the `dataflow` analysis stuff. 

https://discordapp.com/channels/442252698964721669/443151243398086667/492702651759329280

### PR DESCRIPTION
**Pull request by jonas-schievink**
Fixes https://github.com/rust-lang/rust/issues/54444

### CODE REVIEW

File: src/librustc_mir_build/lints.rs [hunk 1]
```diff
@@ -1,138 +1,159 @@
...
+            let terminator = body[bb].terminator();
+            let relevant_successors = match &terminator.kind {
+                TerminatorKind::Call { destination: Some((_, dest)), .. } => {
+                    Some(dest).into_iter().chain(&[])
                 }
-                TerminatorKind::Abort | TerminatorKind::Return => {
-                    //found a path!
-                    reached_exit_without_self_call = true;
-                    break;
+                TerminatorKind::Call { destination: None, .. } => None.into_iter().chain(&[]),
```
Comment (Centril) [inline src/librustc_mir_build/lints.rs:53]: Merge with the arm that includes `::Goto`?

File: src/librustc_mir_build/lints.rs [hunk 2]
```diff
@@ -1,138 +1,159 @@
...
+                TerminatorKind::SwitchInt { targets, .. } => None.into_iter().chain(targets),
+                TerminatorKind::Goto { target }
+                | TerminatorKind::Drop { target, .. }
+                | TerminatorKind::DropAndReplace { target, .. }
+                | TerminatorKind::Assert { target, .. } => Some(target).into_iter().chain(&[]),
+                TerminatorKind::Yield { .. } | TerminatorKind::GeneratorDrop => {
```
Comment (Centril) [inline src/librustc_mir_build/lints.rs:59]: And here also?

File: src/librustc_mir_build/lints.rs [hunk 3]
```diff
@@ -1,138 +1,159 @@
...
         });
     }
 }
+
+/// Finds blocks with `Call` terminators that would end up calling back into the same method.
+fn find_blocks_calling_self<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &Body<'tcx>,
+    def_id: DefId,
+) -> BitSet<BasicBlock> {
```
Comment (Centril) [inline src/librustc_mir_build/lints.rs:122]: Could use more commentary in this function.

Comment (jonas-schievink) [reply src/librustc_mir_build/lints.rs:120]: I tried, but I'm not sure I fully understand it

**Comment by ecstatic-morse** [comment]
I think this is more complex than it could be. To fix the first issue in #54444, you can suppress the lint if `Body::is_cfg_cyclic` returns `true`. This treats any loop as possibly infinite, which I believe this does as well? For the second issue, you just need to properly handle terminators that could unwind by traversing all reachable basic blocks and looking for unwind edges as well as returns. Am I missing something here?

**Comment by jonas-schievink** [comment]
I think checking for cycles in the cfg is not enough - we do want to lint if there's a call before the cycle, but not if it comes after.

**Comment by ecstatic-morse** [comment]
In that case you would need to check for back-edges during the traversal, not as a separate step. There's a [DFS framework with decent documentation](https://github.com/rust-lang/rust/blob/b543afca9b90ad6e4689b6d613ab51f9d3ba15e7/src/librustc_data_structures/graph/iterate/mod.rs#L145) that can differentiate between a cycle and a node with multiple predecessors. You could almost just use it directly, but you need to ignore successors of recursive `Call` terminators, and that framework works on generic graphs.

Adding an `ignore_successors` method to https://github.com/rust-lang/rust/blob/b543afca9b90ad6e4689b6d613ab51f9d3ba15e7/src/librustc_data_structures/graph/iterate/mod.rs#L237 and checking it before pushing successors onto the stack would work: https://github.com/rust-lang/rust/blob/b543afca9b90ad6e4689b6d613ab51f9d3ba15e7/src/librustc_data_structures/graph/iterate/mod.rs#L204-L213.

I do think it's worth implementing this using forward traversal since we can avoid computing the predecessor graph, but if you're happy with your approach, we could merge it with some more documentation. WDYT?

**Comment by jonas-schievink** [comment]
Oh that's neat, didn't know there was an entire framework for that.

> I do think it's worth implementing this using forward traversal since we can avoid computing the predecessor graph

It was already getting computed anyways. I'm not sure what's using it though.

**Comment by jonas-schievink** [comment]
I guess this makes sense since the query returns `&'tcx Steal<BodyAndCache<'tcx>>`, so not having the predecessors updated means the users of the query would have to do it.

**Comment by jonas-schievink** [comment]
@ecstatic-morse okay, added some more comments. Let me know if there's anything else.

File: src/librustc_mir_build/lints.rs [hunk 4]
```diff
@@ -28,7 +28,12 @@ fn check_fn_for_unconditional_recursion<'tcx>(
     def_id: DefId,
 ) {
     let self_calls = find_blocks_calling_self(tcx, &body, def_id);
+
+    // Stores a list of `Span`s for every basic block. Those are the spans of `Call` terminators
+    // where we know that one of them will definitely be reached.
```
Comment (ecstatic-morse) [inline src/librustc_mir_build/lints.rs:33]: If I understand right, when the vector is empty, it means that we **do not** always end up at a recursive call from this basic block. When it has one or more spans, it means we **do** always end up at a recursive call from this basic block, and the vector contains the spans of all the recursive calls we could possibly reach. Is this correct? If so, could you make this a bit clearer in the comment or by giving a better name to `results`?

File: src/librustc_mir_build/lints.rs [hunk 5]
```diff
@@ -1,138 +1,174 @@
...
+        let locations = if self_calls.contains(bb) {
+            // `bb` *is* a self-call.
+            // We don't look at successors here because they are irrelevant here and we don't want
+            // to lint them (eg. `f(); f()` should only lint the first call).
+            vec![bb]
+        } else {
+            // If *all* successors of `bb` lead to a self-call, emit notes at all of their
+            // locations.
+
+            // Converging successors without unwind paths.
```
Comment (ecstatic-morse) [inline src/librustc_mir_build/lints.rs:54]: I'm confused what "converging" means here. This is just `successors` but with the `unwind` edges removed, right?

**Review by ecstatic-morse — COMMENTED** [review]
I asked for some clarification/changes in a few places that confused me initially, but I think this is good overall. I reserve the right to try rewriting this to use only forward traversal :grin:.

r=me with comments (mine and @Centril's) addressed.

File: src/librustc_mir_build/lints.rs (hunk 4 above)
Comment (jonas-schievink) [reply src/librustc_mir_build/lints.rs]: Yeah, that's right. I've updated the comment.

**Pull request merged**